File: src/main/java/com/mrpowergamerbr/loritta/utils/temmieyoutube/TemmieYouTube.java
package com.mrpowergamerbr.loritta.utils.temmieyoutube;

import com.github.kevinsawicki.http.HttpRequest;
import com.mrpowergamerbr.loritta.Loritta;
import com.mrpowergamerbr.loritta.utils.temmieyoutube.response.SearchResponse;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.StringUtils;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Map;

@Getter
@Setter
public class TemmieYouTube {
	private String apiKey;
	
	public TemmieYouTube(String apiKey) {
		this.apiKey = apiKey;
	}
	
	public SearchResponse searchOnYouTube(String searchQuery) {
		Map<String, Object> params = new HashMap<String, Object>();
		params.put("part", "snippet");
		params.put("q", searchQuery);
		params.put("key", apiKey);
		HttpRequest req = HttpRequest.get("https://www.googleapis.com/youtube/v3/search?" + buildQuery(params));

		String body = req.body();

		return Loritta.getGson().fromJson(body, SearchResponse.class);
	}

	public void searchOnGoogle(String searchQuery) {
		Map<String, Object> params = new HashMap<String, Object>();
		params.put("q", searchQuery);
		params.put("cx", "017576662512468239146:omuauf_lfve"); // wtf is this?
		params.put("key", apiKey);
		HttpRequest req = HttpRequest.get("https://www.googleapis.com/customsearch/v1?" + buildQuery(params));

		String body = req.body();
		
		System.out.println(body);
		// return Loritta.getGson().fromJson(body, SearchResponse.class);
	}
	
	private String buildQuery(Map<String, Object> params) {
		String[] query = new String[params.size()];
		int index = 0;
		for (String key : params.keySet()) {
			String val = String.valueOf(params.get(key) != null ? params.get(key) : "");
			try {
				val = URLEncoder.encode(val, "UTF-8");
			} catch (UnsupportedEncodingException e) {
			}
			query[index++] = key+"="+val;
		}

		return StringUtils.join(query, "&");
	}
}

File: src/main/java/com/mrpowergamerbr/loritta/utils/temmieyoutube/TemmieYouTube.kt
package com.mrpowergamerbr.loritta.utils.temmieyoutube

import com.github.kevinsawicki.http.HttpRequest
import com.mrpowergamerbr.loritta.Loritta
import org.apache.commons.lang3.StringUtils
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.util.*

class TemmieYouTube(private val apiKey: String) {
	/**
	 * Pesquisa algo no YouTube
	 *
	 * @return O resultado da pesquisa
	 */
	fun searchOnYouTube(searchQuery: String): SearchResponse {
		val params = HashMap<String, Any>()
		params.put("part", "snippet")
		params.put("q", searchQuery)
		params.put("key", apiKey)
		val req = HttpRequest.get("https://www.googleapis.com/youtube/v3/search?" + buildQuery(params))

		val body = req.body()

		return Loritta.gson.fromJson(body, SearchResponse::class.java)
	}

	private fun buildQuery(params: Map<String, Any>): String {
		val query = arrayOfNulls<String>(params.size)
		var index = 0
		for (key in params.keys) {
			var `val` = (if (params[key] != null) params[key] else "").toString()
			try {
				`val` = URLEncoder.encode(`val`, "UTF-8")
			} catch (e: UnsupportedEncodingException) {
			}

			query[index++] = key + "=" + `val`
		}

		return StringUtils.join(query, "&")
	}
}

data class SearchResponse(
		val kind: String,
		val etag: String,
		val nextPageToken: String,
		val regionCode: String,
		val pageInfo: String,
		val items: List<YouTubeItem>)

data class YouTubeItem(
		val kind: String,
		val etag: String,
		val id: YouTubeId,
		val snippet: Snippet)

data class Thumbnails(
		val default: ThumbnailInfo,
		val medium: ThumbnailInfo,
		val high: ThumbnailInfo)

data class Snippet(
		val publishedAt: String,
		val channelId: String,
		val title: String,
		val description: String,
		val thumbnails: Thumbnails,
		val channelTitle: String,
		val liveBroadcastContent: String)

data class PageInfo(
		val totalResults: Int,
		val resultsPerPage: Int)

data class ThumbnailInfo(
		val url: String,
		val width: Int,
		val height: Int)

data class YouTubeId(
		val kind: String,
		val videoId: String)